File: 0238-Product-of-Array-Except-Self/Solution.py
# Completed on: 9/15/2024
# Submission: https://leetcode.com/problems/product-of-array-except-self/submissions/1391596908/
# Time Complexity: O(n)
# Space Complexity: O(n) 

class Solution:
    def productExceptSelf(self, nums: list[int]) -> list[int]:

        left = []
        right = []
        curr = 1
        for i in nums:
            curr *=i
            left.append(curr)
        
        i = len(nums) -1
        curr = 1
        while i >=0:
            curr *=nums[i]
            right.append(curr)
            i -=1
        right = list(reversed(right))        

        discluded = [0]*len(nums)
        for i in range(len(nums)):
            if i == 0: 
                discluded[i] = right[1]
            elif i == len(nums) -1:
                discluded[i] = left[-2]
            else:
                discluded[i] = left[i- 1] * right[i +1]
        return discluded

File: 0238-Product-of-Array-Except-Self/Solution.cpp
// Completed on: 7/13/2023
// Submission: https://leetcode.com/problems/product-of-array-except-self/submissions/993750291/
// Time Complexity: O(n^2)
// Space Complexity: O(n) 
//   k : Number of distinct elements


#include <vector>
using namespace std;
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        vector<int> output(nums.size());
        output[0] = 1;
        int right = 1;
        for (int i = 1; i < nums.size(); i++ )
        {
            output[i] = output[i - 1] * nums[i - 1];
        }
        for(int i = nums.size() - 1; i >=0; i--)
        {
            output[i] *= right;
            right *= nums[i];
        }
    return output;
    }
    
};
